imap_processing/mag/l1d/mag_l1d_data.py

```
@@ -0,0 +1,572 @@
+
+        return spin_offsets
+
+    def generate_spin_offset_dataset(self) -> xr.Dataset | None:
+        """
+        Output the spin offsets file as a dataset.
+
+        Returns
+        -------
+        xr.Dataset | None
+            The spin offsets dataset. This function can be used to control the output
+            structure of the offsets dataset ancillary file, without affecting how
+            the offsets are used inside the class.
+        """
+        return self.spin_offsets
+
+    @staticmethod
+    def apply_spin_offsets(
+        spin_offsets: xr.Dataset,
+        epoch: np.ndarray,
+        vectors: np.ndarray,
+        spin_average_application_factor: np.float64,
+    ) -> np.ndarray:
+        """
+        Apply the spin offsets to the input vectors.
+
+        This uses the spin offsets calculated by `calculate_spin_offsets` (or passed in
+        to the class in burst mode) to apply the offsets to the input vectors.
+
+        For each vector, we take the nearest offset, multiply it by the
+        spin_average_application_factor calibration value, and subtract the offset from
+        the appropriate axis.
+
+        These spin offsets act as an automatic smoothing effect on the data over each
+        series of spins.
+
+        Parameters
+        ----------
+        spin_offsets : xr.Dataset
+            The spin offsets dataset.
+        epoch : np.ndarray
+            The epoch values for the input vectors, shape (N,).
+        vectors : np.ndarray
+            The input vectors to apply offsets to, shape (N, 3). Can be Mago, magi,
+            burst or norm. The same offsets file is applied to all.
+        spin_average_application_factor : np.float64
+            The spin average application factor from the configuration file.
+
+        Returns
+        -------
+        np.ndarray
+            The output vectors with spin offsets applied, shape (N, 3).
+        """
+        if spin_offsets is None:
```

spin_offsets can be None when apply_gradiometry is False or in modes other than NORM. Are you certain you need to throw an error here? If so, that's fine. Just wanted to double check.

---

Yes, if we're applying offsets we need to have spin_offsets. if we're in BURST mode we need to pass the spin offsets in, so they shouldn't be none.